File: game/game_engine.py
import random

from .ability_system import AbilitySystem


class GameEngine:

    def __init__(
        self,
        player_a,
        player_b,
        ability_system=None,
    ):
        self.players = {
            player_a.id: player_a,
            player_b.id: player_b,
        }

        self.player_order = [
            player_a.id,
            player_b.id,
        ]

        self.abilities = (
            ability_system
            if ability_system is not None
            else AbilitySystem()
        )

        self.turn_number = 0
        self.active_player_id = None
        self.winner_id = None
        self.is_over = False

        self.log = []

    def _emit(self, event):
        self.log.append({
            "turn": self.turn_number,
            **event,
        })

    def get_opponent_id(self, player_id):
        for player_id_in_order in (
            self.player_order
        ):
            if player_id_in_order != player_id:
                return player_id_in_order

        raise ValueError(
            "Could not find opponent."
        )

    def start_game(self):
        first = random.choice(
            self.player_order
        )

        self.active_player_id = first
        self.turn_number = 1

        self._emit({
            "type": "GAME_START",
            "firstPlayerId": first,
        })

        return first

    def _require_active_player(
        self,
        player_id,
    ):
        if self.is_over:
            raise ValueError(
                "Game is already over."
            )

        if player_id != self.active_player_id:
            raise ValueError(
                f"It's not {player_id}'s turn."
            )

    def perform_attack(
        self,
        player_id,
        attacker_slot_index,
        target_slot_index,
        use_super=False,
    ):
        self._require_active_player(
            player_id
        )

        attacker_player = self.players[
            player_id
        ]

        defender_id = self.get_opponent_id(
            player_id
        )

        defender_player = self.players[
            defender_id
        ]

        if not (
            0
            <= attacker_slot_index
            < len(
                attacker_player.active_slots
            )
        ):
            raise ValueError(
                "Invalid attacker slot."
            )

        if not (
            0
            <= target_slot_index
            < len(
                defender_player.active_slots
            )
        ):
            raise ValueError(
                "Invalid target slot."
            )

        attacker = (
            attacker_player.active_slots[
                attacker_slot_index
            ]
        )

        target = (
            defender_player.active_slots[
                target_slot_index
            ]
        )

        if (
            attacker is None
            or not attacker.is_alive
        ):
            raise ValueError(
                "No living attacker in that slot."
            )

        if (
            target is None
            or not target.is_alive
        ):
            raise ValueError(
                "No living target in that slot."
            )

        if (
            use_super
            and not attacker.super_unlocked
        ):
            raise ValueError(
                "That card's super isn't "
                "unlocked yet."
            )

        # Combat reveals both cards.

        attacker.reveal()
        target.reveal()

        ability_id = (
            attacker.template.super_ability_id
            if use_super
            else attacker.template.basic_attack_id
        )

        result = self.abilities.resolve(
            ability_id,
            {
                "engine": self,
                "source": attacker,
                "target": target,

                "all_allies": [
                    card
                    for card
                    in attacker_player.active_slots
                    if card is not None
                ],

                "all_enemies": [
                    card
                    for card
                    in defender_player.active_slots
                    if card is not None
                ],
            },
        )

        # First successful attack unlocks super.

        attacker.mark_acted()

        self._emit({
            "type": (
                "SUPER_ATTACK"
                if use_super
                else "BASIC_ATTACK"
            ),

            "playerId": player_id,

            "attackerId":
                attacker.instance_id,

            "targetId":
                target.instance_id,

            "attackerName":
                attacker.template.name,

            "targetName":
                target.template.name,

            "result": result,
        })

        self._check_for_defeats(
            attacker_player
        )

        self._check_for_defeats(
            defender_player
        )

        self._check_win_condition()

        if not self.is_over:
            self._advance_turn()

        return result

    def perform_swap(
        self,
        player_id,
        reserve_index,
        active_slot_index,
    ):
        self._require_active_player(
            player_id
        )

        player = self.players[
            player_id
        ]

        card = player.swap_into_active(
            reserve_index,
            active_slot_index,
        )

        self._emit({
            "type": "SWAP",
            "playerId": player_id,
            "cardId": card.instance_id,
            "cardName": card.template.name,
            "activeSlotIndex":
                active_slot_index,
        })

        self._advance_turn()

        return card

    def _check_for_defeats(
        self,
        player,
    ):
        cards_to_check = (
            list(player.active_slots)
            + list(player.reserve)
        )

        for card in cards_to_check:
            if (
                card is not None
                and card.is_defeated
                and card.zone != "graveyard"
            ):
                player.handle_defeat(card)

                self._emit({
                    "type": "DEFEATED",
                    "playerId": player.id,
                    "cardId": card.instance_id,
                    "cardName":
                        card.template.name,
                })

    def _check_win_condition(self):
        for player_id in (
            self.player_order
        ):
            if self.players[
                player_id
            ].has_lost:

                self.is_over = True

                self.winner_id = (
                    self.get_opponent_id(
                        player_id
                    )
                )

                self._emit({
                    "type": "GAME_OVER",
                    "winnerId":
                        self.winner_id,
                    "loserId":
                        player_id,
                })

                return

    def _advance_turn(self):
        self.active_player_id = (
            self.get_opponent_id(
                self.active_player_id
            )
        )

        self.turn_number += 1

    def get_state_for(
        self,
        player_id,
    ):
        if player_id not in self.players:
            raise ValueError(
                "Unknown player."
            )

        opponent_id = (
            self.get_opponent_id(
                player_id
            )
        )

        me = self.players[player_id]
        opponent = self.players[
            opponent_id
        ]

        return {
            "turnNumber":
                self.turn_number,

            "activePlayerId":
                self.active_player_id,

            "isOver":
                self.is_over,

            "winnerId":
                self.winner_id,

            "you": {
                "id": me.id,
                "name": me.name,

                "active": [
                    card.to_owner_view()
                    if card is not None
                    else None
                    for card in me.active_slots
                ],

                "reserve": [
                    card.to_owner_view()
                    for card in me.reserve
                ],
            },

            "opponent": {
                "id": opponent.id,
                "name": opponent.name,

                "active": [
                    card.to_public_view()
                    if card is not None
                    else None
                    for card
                    in opponent.active_slots
                ],

                "reserveCount":
                    len(opponent.reserve),
            },

            "log": self.log[-30:],
        }

File: src/GameEngine.js
import { AbilitySystem } from "./AbilitySystem.js";

/**
 * GameEngine
 * ----------
 * Owns the authoritative game state and rules. Nothing here knows about
 * rendering — feed it player actions, read back state/events, and hook a
 * UI (or CLI, or a bot) up to it however you like.
 *
 * Core rules encoded here (from your description):
 *  - 5 active slots + 2 reserve per player.
 *  - Coin flip decides who goes first.
 *  - Cards are face-down (unrevealed) until they attack or are attacked;
 *    both attacker and target get revealed the moment combat happens.
 *  - A card's super unlocks the first time it successfully attacks; it's
 *    usable (not forced) any time after that, including the same turn
 *    it unlocks — up to you to change if you want a one-turn delay.
 *  - A player loses when they have no living cards in active or reserve.
 *
 * Anything ability-specific is delegated to AbilitySystem so you can
 * redesign brawler kits without touching this file.
 */

export const ActionType = Object.freeze({
  BASIC_ATTACK: "BASIC_ATTACK",
  SUPER_ATTACK: "SUPER_ATTACK",
  SWAP: "SWAP",
});

export class GameEngine {
  /**
   * @param {import("./Player.js").Player} playerA
   * @param {import("./Player.js").Player} playerB
   * @param {AbilitySystem} [abilitySystem]
   */
  constructor(playerA, playerB, abilitySystem = new AbilitySystem()) {
    this.players = { [playerA.id]: playerA, [playerB.id]: playerB };
    this.playerOrder = [playerA.id, playerB.id];
    this.abilities = abilitySystem;

    this.turnNumber = 0;
    this.activePlayerId = null;
    this.winnerId = null;
    this.isOver = false;
    this.log = []; // append-only event log, useful for UI/replay/debugging
  }

  _emit(event) {
    this.log.push({ turn: this.turnNumber, ...event });
  }

  getOpponentId(playerId) {
    return this.playerOrder.find((id) => id !== playerId);
  }

  /** Flip a coin to decide who goes first, then start turn 1. */
  startGame() {
    const first = Math.random() < 0.5 ? this.playerOrder[0] : this.playerOrder[1];
    this.activePlayerId = first;
    this.turnNumber = 1;
    this._emit({ type: "GAME_START", firstPlayerId: first });
    return first;
  }

  _requireActivePlayer(playerId) {
    if (this.isOver) throw new Error("Game is already over.");
    if (playerId !== this.activePlayerId) {
      throw new Error(`It's not ${playerId}'s turn.`);
    }
  }

  /**
   * @param {string} playerId
   * @param {number} attackerSlotIndex - index into the attacker's activeSlots
   * @param {number} targetSlotIndex - index into the opponent's activeSlots
   * @param {boolean} [useSuper=false]
   */
  performAttack(playerId, attackerSlotIndex, targetSlotIndex, useSuper = false) {
    this._requireActivePlayer(playerId);

    const attackerPlayer = this.players[playerId];
    const defenderId = this.getOpponentId(playerId);
    const defenderPlayer = this.players[defenderId];

    const attacker = attackerPlayer.activeSlots[attackerSlotIndex];
    const target = defenderPlayer.activeSlots[targetSlotIndex];

    if (!attacker || !attacker.isAlive) {
      throw new Error("No living attacker in that slot.");
    }
    if (!target || !target.isAlive) {
      throw new Error("No living target in that slot.");
    }
    if (useSuper && !attacker.superUnlocked) {
      throw new Error("That card's super isn't unlocked yet.");
    }

    // Reveal rule: both cards involved in combat are revealed, win or lose.
    attacker.reveal();
    target.reveal();

    const abilityId = useSuper ? attacker.template.superAbilityId : attacker.template.basicAttackId;
    const result = this.abilities.resolve(abilityId, {
      engine: this,
      source: attacker,
      target,
      allAllies: attackerPlayer.activeSlots.filter(Boolean),
      allEnemies: defenderPlayer.activeSlots.filter(Boolean),
    });

    attacker.markActed();

    this._emit({
      type: useSuper ? "SUPER_ATTACK" : "BASIC_ATTACK",
      playerId,
      attackerId: attacker.instanceId,
      targetId: target?.instanceId,
      result,
    });

    // Sweep for defeats caused by this action (covers heals-only abilities
    // doing nothing here, and damage abilities knocking out a card).
    this._checkForDefeats(attackerPlayer);
    this._checkForDefeats(defenderPlayer);

    this._checkWinCondition();
    if (!this.isOver) this._advanceTurn();

    return result;
  }

  /**
   * Bring a reserve card into an empty active slot. Costs the turn, same
   * as an attack — change `_advanceTurn` call below if you'd rather let
   * swaps be free/instant.
   */
  performSwap(playerId, reserveIndex, activeSlotIndex) {
    this._requireActivePlayer(playerId);
    const player = this.players[playerId];
    const card = player.swapIntoActive(reserveIndex, activeSlotIndex);

    this._emit({
      type: "SWAP",
      playerId,
      cardId: card.instanceId,
      activeSlotIndex,
    });

    this._advanceTurn();
  }

  _checkForDefeats(player) {
    for (const card of [...player.activeSlots, ...player.reserve]) {
      if (card && card.isDefeated && card.zone !== "graveyard") {
        player.handleDefeat(card);
        this._emit({ type: "DEFEATED", playerId: player.id, cardId: card.instanceId });
      }
    }
  }

  _checkWinCondition() {
    for (const id of this.playerOrder) {
      if (this.players[id].hasLost) {
        this.isOver = true;
        this.winnerId = this.getOpponentId(id);
        this._emit({ type: "GAME_OVER", winnerId: this.winnerId, loserId: id });
        return;
      }
    }
  }

  _advanceTurn() {
    this.activePlayerId = this.getOpponentId(this.activePlayerId);
    this.turnNumber += 1;
  }

  /**
   * Returns game state from one player's point of view: their own cards
   * fully visible, the opponent's unrevealed cards masked as face-down.
   */
  getStateFor(playerId) {
    const opponentId = this.getOpponentId(playerId);
    const me = this.players[playerId];
    const opp = this.players[opponentId];

    return {
      turnNumber: this.turnNumber,
      activePlayerId: this.activePlayerId,
      isOver: this.isOver,
      winnerId: this.winnerId,
      you: {
        id: me.id,
        name: me.name,
        active: me.activeSlots.map((c) => (c ? c.toOwnerView() : null)),
        reserve: me.reserve.map((c) => c.toOwnerView()),
      },
      opponent: {
        id: opp.id,
        name: opp.name,
        active: opp.activeSlots.map((c) => (c ? c.toPublicView() : null)),
        reserveCount: opp.reserve.length,
      },
    };
  }
}
